File: backend/app/api/mypage.py
from fastapi import APIRouter, Depends, HTTPException, Response
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.security import ACCESS_TOKEN_COOKIE_NAME, get_current_user
from app.models.user import User
from app.schemas.mypage import (
    AccountDeleteRequest,
    AccountDeleteResponse,
    MyPageResponse,
    PasswordChangeRequest,
    PasswordChangeResponse,
    StoreUpdateRequest,
    StoreUpdateResponse,
)
from app.services.mypage import get_mypage, change_password, update_store, delete_account


router = APIRouter(
    prefix="/mypage",
    tags=["MyPage"],
)

#마이페이지 조회
@router.get(
    "",
    response_model=MyPageResponse,
)
def mypage(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    result = get_mypage(db, current_user.user_id)

    if result is None:
        raise HTTPException(
            status_code=404,
            detail="마이페이지 정보를 찾을 수 없습니다.",
        )

    user, store = result

    return {
        "store": {
            "business_name": store.business_name,
            "industry_name": store.industry_name,
            "business_address": store.business_address,
            "open_date": store.open_date,
        },
        "business_verification": {
            "business_reg_no": user.business_reg_no,
            "representative_name": user.representative_name,
            "industry_name": store.industry_name,
        },
        "account": {
            "login_id": user.business_reg_no,
            "phone_number": user.phone_number,
        },
    }

#비밀번호 변경
@router.patch(
    "/password",
    response_model=PasswordChangeResponse,
)
def update_password(
    request: PasswordChangeRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        change_password(
            db=db,
            user=current_user,
            current_password=request.current_password,
            new_password=request.new_password,
        )

        return {
            "message": "비밀번호가 변경되었습니다."
        }

    except ValueError as e:
        raise HTTPException(
            status_code=400,
            detail=str(e),
        )
        
# 가게 정보 수정
@router.patch(
    "/store",
    response_model=StoreUpdateResponse,
)
def update_store_info(
    request: StoreUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        store = update_store(
            db=db,
            user_id=current_user.user_id,
            business_name=request.business_name,
            industry_name=request.industry_name,
            business_address=request.business_address,
            open_date=request.open_date,
        )

        return {
            "message": "가게 정보가 수정되었습니다.",
            "store": {
                "business_name": store.business_name,
                "industry_name": store.industry_name,
                "business_address": store.business_address,
                "open_date": store.open_date,
            },
        }

    except ValueError as e:
        raise HTTPException(
            status_code=404,
            detail=str(e),
        )

# 회원 탈퇴
@router.delete(
    "/account",
    response_model=AccountDeleteResponse,
)
def delete_my_account(
    request: AccountDeleteRequest,
    response: Response,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        delete_account(db, current_user, request.password)
    except ValueError as e:
        raise HTTPException(
            status_code=400,
            detail=str(e),
        )

    # 탈퇴한 계정의 세션 쿠키도 지운다.
    response.delete_cookie(key=ACCESS_TOKEN_COOKIE_NAME, path="/")

    return {
        "message": "회원 탈퇴가 완료되었습니다."
    }

File: frontend/src/types/mypage.ts
/**
 * GET /api/v1/mypage · PATCH /mypage/store · PATCH /mypage/password · DELETE /mypage/account 요청·응답.
 * mypage는 envelope 없이 response_model 그대로(snake_case) 내려오고, 실패는 `{ detail }`입니다.
 */

export interface MypageStoreResponse {
  business_name: string
  industry_name: string
  business_address: string
  /** 개업일 YYYY-MM-DD */
  open_date: string
  /** 업태. 백엔드 추가 예정이라 없을 수 있습니다. */
  business_type?: string
  /** 종목. 백엔드 추가 예정이라 없을 수 있습니다. */
  business_category?: string
}

export interface MypageVerificationResponse {
  /** 하이픈 없는 10자리 */
  business_reg_no: string
  representative_name: string
  industry_name: string
  business_type?: string
  business_category?: string
}

export interface MypageAccountResponse {
  /** 로그인 아이디. 사업자등록번호와 같습니다. */
  login_id: string
  /** 하이픈 없는 11자리 */
  phone_number: string
}

export interface MypageResponse {
  store: MypageStoreResponse
  business_verification: MypageVerificationResponse
  account: MypageAccountResponse
}

export interface StoreUpdateRequest {
  business_name: string
  industry_name: string
  business_address: string
  open_date: string
  business_type?: string
  business_category?: string
}

export interface StoreUpdateResponse {
  message: string
  store: MypageStoreResponse
}

export interface PasswordChangeRequest {
  current_password: string
  new_password: string
}

export interface AccountDeleteRequest {
  password: string
}
